### ISSUE TITLE
Drupal 10 compatibility.

### ISSUE THREAD

**Issue by dpagini** [apigee/apigee-edge-drupal#761]
## Is your feature request related to a problem? Please describe.
Release a module version that is compatible with Drupal 10.

## Describe the solution you would like
Need to have this module ship a D10/PHP8.1 compatible version before we can update our site to Drupal 10.

## Describe alternatives you have considered
Not using the Apigee suite of modules?

## Additional context 
N/A

**Comment by VladimirAus** [comment]
Running `./vendor/bin/drush -l default -y upgrade_status:analyze apigee_edge`

### Result

```

================================================================================
Apigee Edge,  2.0.6
Scanned on Sat, 07/01/2023 - 19:36

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_teams/src/Controller/TeamMembers
List.php

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 118  Relying on entity queries to check access by default is     
                    deprecated in drupal:9.2.0 and an error will be thrown from 
                    drupal:10.0.0. Call                                         
                    \Drupal\Core\Entity\Query\QueryInterface::accessCheck() with
                    TRUE or FALSE to specify whether access should be checked.  
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_teams/src/Entity/Form/TeamForm.p
hp

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 156  Relying on entity queries to check access by default is     
                    deprecated in drupal:9.2.0 and an error will be thrown from 
                    drupal:10.0.0. Call                                         
                    \Drupal\Core\Entity\Query\QueryInterface::accessCheck() with
                    TRUE or FALSE to specify whether access should be checked.  
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_teams/src/Entity/Storage/TeamApp
Storage.php

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 38   Class Drupal\apigee_edge_teams\Entity\Storage\TeamAppStorage
                    extends @internal class                                     
                    Drupal\apigee_edge\Entity\Storage\AppStorage.               
--------------------------------------------------------------------------------

FILE: modules/contrib/apigee_edge/apigee_edge.libraries.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    The 'apigee_edge.analytics' library is depending on a       
                    deprecated library. The core/jquery.once asset library is   
                    deprecated in Drupal 9.3.0 and will be removed in Drupal    
                    10.0.0. Use the core/once library instead. See              
                    https://www.drupal.org/node/3158256                         
--------------------------------------------------------------------------------
Check manually 0    The 'apigee_edge.app_view' library is depending on a        
                    deprecated library. The core/jquery.once asset library is   
                    deprecated in Drupal 9.3.0 and will be removed in Drupal    
                    10.0.0. Use the core/once library instead. See              
                    https://www.drupal.org/node/3158256                         
--------------------------------------------------------------------------------
Check manually 0    The 'apiproduct_access_admin' library is depending on a     
                    deprecated library. The core/jquery.once asset library is   
                    deprecated in Drupal 9.3.0 and will be removed in Drupal    
                    10.0.0. Use the core/once library instead. See              
                    https://www.drupal.org/node/3158256                         
--------------------------------------------------------------------------------

FILE: modules/contrib/apigee_edge/apigee_edge.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8.7.7 || ^9 is not      
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_debug/apigee_edge_debug.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_actions/apigee_edge_actions.info
.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_actions/modules/apigee_edge_acti
ons_examples/apigee_edge_actions_examples.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_actions/modules/apigee_edge_acti
ons_debug/apigee_edge_actions_debug.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_teams/apigee_edge_teams.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE:
modules/contrib/apigee_edge/modules/apigee_edge_apiproduct_rbac/apigee_edge_apip
roduct_rbac.info.yml

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    Value of core_version_requirement: ^8 || ^9 is not          
                    compatible with the next major version of Drupal core. See  
                    https://drupal.org/node/3070687.                            
--------------------------------------------------------------------------------

FILE: modules/contrib/apigee_edge/composer.json

STATUS         LINE                           MESSAGE                           
--------------------------------------------------------------------------------
Check manually 0    The drupal/core requirement is not compatible with the next 
                    major version of Drupal. Either remove it or update it to be
                    compatible. See                                             
                    https://drupal.org/node/2514612#s-drupal-9-compatibility.   
--------------------------------------------------------------------------------


```

**Comment by phdhiren** [comment]
@dpagini, @VladimirAus,

Thank you for identifying the issues and sharing the information. I can create individual separate issues for each one of these changes. I will use the label "Drupal 10 Preparation" for this.

**Comment by mxr576** [comment]
@giteshk Hi, is there a timeline that you can share publicly for adding D10 compatibility to this module? Thanks in advance.

**Comment by giteshk** [comment]
We have started work on this, tracking to end of Q2 2023.

You will start seeing issues being created in the next 2 weeks. 
- [Apigee Edge Drupal module ](https://github.com/apigee/apigee-edge-drupal/issues?q=is%3Aissue+label%3A%22Drupal+10+Preparation%22+is%3Aopen)
- [Apigee Edge Client Library for PHP](https://github.com/apigee/apigee-client-php/issues?q=is%3Aissue+is%3Aopen+label%3A%22Drupal+10+Preparation%22)

Feel free to add any tickets with "Drupal 10 Preparation" label.
We love contributions from the community !

**Comment by mxr576** [comment]
Thanks for the quick reply Gitesh!

> We have started work on this, tracking to end of Q2 2023.


Q2 has just started, Drupal 10.0.0 is out since December 14, 2022. Drupal 9.5's support end on November 1, 2023.

These dates means  to me that Drupal 10.0.x should be fairly stable now and if companies do not want to wait until the final call to upgrade their sites to security covered _new major version_ (that also requires PHP 8.1) they should do that either in Q2 or Q3. 
However, the summer holidays always interrupt Q3... so my gut says it is better to start that now.

In other words, can the Apigee team make this a higher priority? 

Regarding contributions, @VladimirAus has already opened the #782. I have spent some time reviewing that back in the day but I see no other feedback on it.

**Comment by janmashat** [comment]
I agree with @mxr576 that customers (and vendors) should be given a chance to test the update in Q2 - because if it's postponed to Q3 then by the time things reach production we could already be in Q4 🙀

**Comment by bdanin** [comment]
To add the chorus here: I help maintain a large site that relies on this module, and it is the only main blocker keeping us from being able to upgrade our multi-site to Drupal 10. For us, it would be ideal if this was already a Drupal 10 release.

**Comment by ChristophWeber** [comment]
I will pile on as well. Any delay past Q2 will put quite a bit of strain on downstream projects. In that regard @giteshk 's update is most welcome, as is @mxr576 's comment and code review.

**Comment by VladimirAus** [comment]
All PRs I did for Apigee Drupal modules in the last 3 years were either stolen by internal Google team (I never got a credit) or forgotten, so no surprises here 🤷‍♂️

**Comment by earth2marsh** [comment]
Hey @VladimirAus , that's pretty strong language! If you're serious about your allegations of "stolen code," I'd be happy to review examples where you have some grievance, and then to address where changes might be necessary. But let's find a way to do that that doesn't pollute this issue, which is meant to address Drupal 10 compatibility?

And thanks, everyone, for your input on how important it is to enable upgrades to Drupal 10. We hear you, and it is our top priority as soon as we wrap up some improvements to how we model app ownership in Apigee X/hybrid. 

If you see opportunities for how we might accelerate that by improving how community contributions get reviewed/accepted, please do offer suggestions (perhaps as a new issue or we could enable discussions for the project?). Note that in general, the bigger the PR and more files it touches, the more complicated the review will be.

**Comment by VladimirAus** [comment]
Hey @earth2marsh. 
You can review [all my PRs that were closed or ignored](https://github.com/apigee/apigee-edge-drupal/pulls?q=is%3Apr+author%3AVladimirAus+) and then the code that is currently in the repo.
Plus I don't see any credits given on [drupal.org](https://www.drupal.org/project/apigee_edge).

**Comment by mxr576** [comment]
How contributions are handled currently and what are the potential improvements in that regard is probably a side track here. Being g that said, please allow me to share my quick feedback on that: I also kept asking for reopening issues that Github Actions auto-closed because there were "no activity" on them. My personal opinion that such feature only works when there is capacity to actively check incoming issues and PRs on a project every day. Without a first response from a maintainer (which is not just an "ACK") no bot should auto-close issues. Also, instead of auto-closing them, first they should move the "stale" status. The Symfony project can be a good inspiration for configuring the bot.

![image](https://user-images.githubusercontent.com/1755573/232034259-c6631d77-2d75-4960-bb78-679c98f3b889.png)
(from https://github.com/apigee/apigee-edge-drupal/pull/602)

**Comment by earth2marsh** [comment]
In order to keep this issue focused on Drupal 10, I enabled Discussions and started a thread where we can discuss what the most effective way to [manage maintainers vs contributors](https://github.com/apigee/apigee-edge-drupal/discussions/837) should be, and another discussion for [suggestions on improving contribution workflows](https://github.com/apigee/apigee-edge-drupal/discussions/838) in the project.

**Comment by janmashat** [comment]
Hi, a couple months have passed here since the last comment...any chance we could get an update on the timeline for releasing a stable D10 version? Some of our customers have begun asking how long they will have for testing the D10 update - and even if Apigee releases a D10 version next week, that leaves 4 months before D9's EOL. That may seem like a lot, but we'll need to prepare, integrate, test, package and deliver before our customers can UAT and schedule deployments. That realistically leaves 2 months for UAT (which is already troubling for some customers) - and that number will quickly decrease over the coming weeks.

**Comment by divya-intelli** [comment]
Hi @janmashat , 
Thank you for your patience. 
We will be releasing Apigee edge module for Drupal 10 next week. Currently we are testing the changes done for Drupal10 in drupal10-prep branch.

**Comment by divya-intelli** [comment]
Hi All, 

We have released Drupal 10 compatible Apigee Edge module version 3.0.0.
@VladimirAus the changes from #782 which are pending in the README.md will be merged in separate PR(issue #880).

**Comment by dpagini** [comment]
@divya-intelli Thanks for the announcement here! I see the new module **is only** compatible with D10 (not D9), which means it can't be tested out until the whole Drupal project moves to D10. Is that a conscience decision by Apigee, or maybe an oversight? If it's compatible with D9, it would be great to see it work with both 9 and 10 versions...

**Comment by divya-intelli** [comment]
Hi @dpagini ,
Yes, 3.x branch is only compatible with Drupal 10. We are maintaining 3.x branch for Drupal 10 due to dependency issue (php-http/guzzle7-adapter) . To test, you will have to upgrade Drupal project to Drupal 10. 
For Drupal 9 you can use 2.x branch.

**Comment by divya-intelli** [comment]
@dpagini ,
We will be supporting 2.x as long as Drupal 9 is supported. Changes will be applied to both the branches.
